core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -354,15 +352,25 @@ private void addMappers(Collection<ObjectMapper> objectMappers, Collection<Field
         mapperService.addMappers(objectMappers, fieldMappers);
     }
 
-    public MergeResult merge(Mapping mapping, boolean simulate, boolean updateAllTypes) {
+    public void merge(Mapping mapping, boolean updateAllTypes) {
         try (ReleasableLock lock = mappingWriteLock.acquire()) {
-            final MergeResult mergeResult = new MergeResult(simulate, updateAllTypes);
+            // first simulate to only check for conflicts
+            MergeResult mergeResult = new MergeResult(true, updateAllTypes);
             this.mapping.merge(mapping, mergeResult);
-            if (simulate == false) {
-                addMappers(mergeResult.getNewObjectMappers(), mergeResult.getNewFieldMappers(), updateAllTypes);
-                refreshSource();
+            if (mergeResult.hasConflicts()) {
+                throw new MergeMappingException(mergeResult.buildConflicts());
             }
-            return mergeResult;
+
+            // then check conflicts with other types
+            // if there is a single type then we are implicitly updating all types
+            mapperService.checkMappersCompatibility(mapping, updateAllTypes || mapperService.types().size() == 1);
+
+            // finally apply the mapping update for real
+            mergeResult = new MergeResult(false, updateAllTypes);
```

Comment:
Notice the difference in the first parameter to MergeResult. This is the "simulate" argument. The first time we don't change anything in the merge, only check for any problems.  Ideally we could move this simulation to something like we have here with check compatibility. I had a branch for a this long ago, but it was a complex change.